### PR TITLE
Add individual store types

### PR DESCRIPTION
**Pull request by LMS007**
- Add types for each store including most of their selectors and actions.
- Not every selector and action is completed, but this covers most of them and serves as a first step to typing all the selectors / actions

### CODE REVIEW

File: src/sidebar/store/modules/session.js [hunk 1]
```diff
@@ -68,6 +77,8 @@ function isFeatureEnabled(state, feature) {
  * Return true if the user's profile has been fetched. This can be used to
  * distinguish the dummy profile returned by `profile()` on startup from a
  * logged-out user profile returned by the server.
+ *
+ * @param {object} state - The application state
```
Comment (robertknight) [inline src/sidebar/store/modules/session.js:81]: Can you remove the `state` param JSDoc for the moment. This changed yesterday such that `state` now refers to the local module state in most cases.

File: src/sidebar/store/modules/toast-messages.js [hunk 2]
```diff
@@ -98,6 +98,20 @@ function hasMessage(state, type, text) {
...
+ * @typedef ToastMessagesStore
+ *
+ * // Actions
+ * @prop {typeof addMessage} addToastMessage
+ * @prop {typeof removeMessage} removeToastMessage
+ * @prop {typeof updateMessage} updateToastMessage
+ *
+ * // Selectors
+ * @prop {(id: string) => (Object|undefined)} getToastMessage
+ * @prop {() => Object[]} getToastMessages
```
Comment (robertknight) [inline src/sidebar/store/modules/toast-messages.js:111]: It would be good to introduce a typedef documenting what properties a toast message has. I don't think we have a suitable typedef anywhere yet. The closest thing to a definition is here: https://github.com/hypothesis/client/blob/f93ea8275b996794db98880efa1e25363b254de7/src/sidebar/services/toast-messenger.js#L62.

File: src/sidebar/store/modules/drafts.js [hunk 3]
```diff
@@ -178,13 +180,29 @@ function getDraftIfNotEmpty(state, annotation) {
 /**
  * Returns a list of draft annotations which have no id.
  *
- * @return {Object[]}
+ * @return {Annotation[]}
```
[the remark this answers is not in the document]
Comment (LMS007) [reply src/sidebar/store/modules/drafts.js:183]: indeed, good to keep that in mind

File: src/sidebar/store/modules/real-time-updates.js [hunk 4]
```diff
@@ -140,7 +140,7 @@ function clearPendingUpdates() {
  * Return added or updated annotations received via the WebSocket
  * which have not been applied to the local state.
  *
- * @return {{[id: string]: Annotation}}
+ * @return {Object<string, boolean>}
```
[the remark this answers is not in the document]
Comment (LMS007) [reply src/sidebar/store/modules/real-time-updates.js:143]: Awe, missed that one. I agree with using the dot notation as a standard.

File: src/sidebar/store/modules/route.js [hunk 5]
```diff
@@ -56,6 +56,17 @@ function routeParams(state) {
...
 
+/**
+ * @typedef RouteStore
+ *
+ * // Actions
+ * @prop {typeof changeRoute} changeRoute
+ *
+ * // Selectors
+ * @prop {() => ?string} route
+ * @prop {() => Object} routeParams
```
[the remark this answers is not in the document]
Comment (LMS007) [reply src/sidebar/store/modules/route.js:67]: okay. I'll change the action (above) too

File: src/sidebar/store/modules/defaults.js [hunk 6]
```diff
@@ -55,6 +55,17 @@ function getDefaults(state) {
...
 }
 
+/**
+ * @typedef DefaultsStore
+ *
+ * // Actions
+ * @prop {typeof setDefault} setDefault
+ *
+ * // Selectors
+ * @prop {(key: string) => string|null} getDefault
```
[the remark this answers is not in the document]
Comment (LMS007) [reply src/sidebar/store/modules/defaults.js:65]: I'm not sure what the syntax would be then for the latter.

File: src/sidebar/store/modules/defaults.js [hunk 7]
```diff
@@ -55,6 +55,17 @@ function getDefaults(state) {
...
+ * @prop {() => (string|null)[]} getDefaults
```
[the remark this answers is not in the document]
Comment (LMS007) [reply src/sidebar/store/modules/defaults.js:66]: Won't it also return `Object.<string,string>)` ?

File: src/sidebar/store/modules/direct-linked.js [hunk 8]
```diff
@@ -140,6 +140,22 @@ function directLinkedGroupFetchFailed(state) {
...
+ *
+ * // Actions
+ * @prop {typeof setDirectLinkedGroupFetchFailed} setDirectLinkedGroupFetchFailed
+ * @prop {typeof setDirectLinkedGroupId} setDirectLinkedGroupId
+ * @prop {typeof setDirectLinkedAnnotationId} setDirectLinkedAnnotationId
+ * @prop {typeof clearDirectLinkedGroupFetchFailed} clearDirectLinkedGroupFetchFailed
+ * @prop {typeof clearDirectLinkedIds} clearDirectLinkedIds
+ *
+ * // Selectors
+ * @prop {() => ?string} directLinkedAnnotationId
```
[the remark this answers is not in the document]
Comment (LMS007) [reply src/sidebar/store/modules/direct-linked.js:154]: Yep, missed that one, this pulled from an older branch I had. Forgot to convert it.

File: src/sidebar/store/modules/session.js (hunk 1 above)
Comment (LMS007) [reply src/sidebar/store/modules/session.js:81]: 👍

File: src/sidebar/services/toast-messenger.js [hunk 9]
```diff
@@ -10,6 +10,14 @@ const MESSAGE_DISPLAY_TIME = 5000;
 // Delay before removing the message entirely (allows animations to complete)
 const MESSAGE_DISMISS_DELAY = 500;
 
+/**
+ * @typedef ToastMessage
```
Comment (robertknight) [inline src/sidebar/services/toast-messenger.js:14]: Can you move this type to the store module? This helps to preserve the overall dependency structure where modules in `services` depend on the `store` but not vice-versa.

File: src/sidebar/store/modules/annotations.js [hunk 10]
```diff
@@ -545,6 +543,35 @@ function savedAnnotations(state) {
...
+ * // Actions
+ * @prop {typeof addAnnotations} addAnnotations
+ * @prop {typeof clearAnnotations} clearAnnotations
+ * @prop {typeof focusAnnotations} focusAnnotations
+ * @prop {typeof hideAnnotation} hideAnnotation
+ * @prop {typeof removeAnnotations} removeAnnotations
+ * @prop {typeof unhideAnnotation} unhideAnnotation
+ * @prop {typeof updateAnchorStatus} updateAnchorStatus
+ * @prop {typeof updateFlagStatus} updateFlagStatus
+
```
Comment (robertknight) [inline src/sidebar/store/modules/annotations.js:558]: There is a blank line here and some trailing space. Maybe install something like [this extension](https://marketplace.visualstudio.com/items?itemName=shardulm94.trailing-spaces#review-details) to highlight trailing whitespace in your IDE?

File: src/sidebar/store/modules/defaults.js [hunk 11]
```diff
@@ -55,6 +55,17 @@ function getDefaults(state) {
...
 
+/**
+ * @typedef DefaultsStore
+ *
+ * // Actions
+ * @prop {typeof setDefault} setDefault
+ *
+ * // Selectors
+ * @prop {(key: string) => string|null} getDefault
+ * @prop {() => Object.<string,null>} getDefaults
```
Comment (robertknight) [inline src/sidebar/store/modules/defaults.js:66]:
```suggestion
 * @prop {() => Object.<string,string|null>} getDefaults
```

The current type means that `getDefaults` values are always `null`, whereas they should be `string|null`.

File: src/sidebar/store/modules/direct-linked.js [hunk 12]
```diff
@@ -140,6 +140,22 @@ function directLinkedGroupFetchFailed(state) {
...
+ * @prop {typeof setDirectLinkedGroupFetchFailed} setDirectLinkedGroupFetchFailed
+ * @prop {typeof setDirectLinkedGroupId} setDirectLinkedGroupId
+ * @prop {typeof setDirectLinkedAnnotationId} setDirectLinkedAnnotationId
+ * @prop {typeof clearDirectLinkedGroupFetchFailed} clearDirectLinkedGroupFetchFailed
+ * @prop {typeof clearDirectLinkedIds} clearDirectLinkedIds
+ *
+ * // Selectors
+ * @prop {() => string|null} directLinkedAnnotationId
+ * @prop {() => boolean} directLinkedGroupFetchFailed
+ * @prop {() => ?string} directLinkedGroupId
```
Comment (robertknight) [inline src/sidebar/store/modules/direct-linked.js:156]:
```suggestion
 * @prop {() => string|null} directLinkedGroupId
```

File: src/sidebar/store/modules/drafts.js [hunk 13]
```diff
@@ -178,13 +180,29 @@ function getDraftIfNotEmpty(state, annotation) {
...
+ * @typedef DraftsStore
+ *
+ * // Actions
+ * @prop {typeof createDraft} createDraft
+ * @prop {typeof deleteNewAndEmptyDrafts} deleteNewAndEmptyDrafts
+ * @prop {typeof discardAllDrafts} discardAllDrafts
+ * @prop {typeof removeDraft} removeDraft
+ *
+ * // Selectors
+ * @prop {() => boolean} countDrafts
```
Comment (robertknight) [inline src/sidebar/store/modules/drafts.js:200]:
```suggestion
 * @prop {() => number} countDrafts
```

File: src/sidebar/store/modules/drafts.js [hunk 14]
```diff
@@ -178,13 +180,29 @@ function getDraftIfNotEmpty(state, annotation) {
...
+ * @prop {(a: Annotation) => Draft|null} getDraft
+ * @prop {(a: Annotation) => Draft|null} getDraftIfNotEmpty
+ * @prop {() => Annotation[]} unsavedAnnotations
```
Comment (robertknight) [inline src/sidebar/store/modules/drafts.js:203]: It is potentially problematic that drafts hold references to `Annotation`s as those copies may not be in sync with the `Annotation` in the `annotations` module. Fixing that is out of scope for this PR though.

File: src/sidebar/store/modules/links.js [hunk 15]
```diff
@@ -22,6 +22,13 @@ function updateLinksAction(newLinks) {
   return { type: 'UPDATE_LINKS', newLinks: newLinks };
 }
 
+/**
+ * @typedef LinksStore
+ *
+ * // Actions
+ * @prop {typeof updateLinks} updateLinksAction
```
Comment (robertknight) [inline src/sidebar/store/modules/links.js:29]:
```suggestion
 * @prop {typeof updateLinks} updateLinks
```

The action creator is renamed in the export as `updateLinks`. The ideal solution here would be to change this store module's structure to match the others, then the action creator function can be named `updateLinks` and the confusion avoided.

File: src/sidebar/store/modules/real-time-updates.js [hunk 16]
```diff
@@ -173,6 +173,20 @@ function hasPendingDeletion(state, id) {
...
+ * @typedef RealTimeUpdatesStore
+ *
+ * // Actions
+ * @prop {typeof receiveRealTimeUpdates} receiveRealTimeUpdates
+ * @prop {typeof clearPendingUpdates} clearPendingUpdates
+ *
+ * // Selectors
+ * @prop {() => boolean} hasPendingDeletion
+ * @prop {() => Object<string, boolean>} pendingDeletions
+ * @prop {() => Object<string, Annotation>} pendingUpdates
```
Comment (robertknight) [inline src/sidebar/store/modules/real-time-updates.js:186]:
```suggestion
 * @prop {() => Object.<string, boolean>} pendingDeletions
 * @prop {() => Object.<string, Annotation>} pendingUpdates
```

Trivial change for stylistic consistency.

File: src/sidebar/store/modules/route.js [hunk 17]
```diff
@@ -56,6 +56,17 @@ function routeParams(state) {
...
 }
 
+/**
+ * @typedef RouteStore
+ *
+ * // Actions
+ * @prop {typeof changeRoute} changeRoute
+ *
+ * // Selectors
+ * @prop {() => ?string} route
```
Comment (robertknight) [inline src/sidebar/store/modules/route.js:66]:
```suggestion
 * @prop {() => string|null} route
```

File: src/sidebar/store/modules/toast-messages.js [hunk 18]
```diff
@@ -1,5 +1,7 @@
 import * as util from '../util';
 
+/** @typedef {import("../../services/toast-messenger").ToastMessage} ToastMessage */
```
Comment (robertknight) [inline src/sidebar/store/modules/toast-messages.js:3]: As noted in `services/toast-messenger`, I'd prefer the typedef to be moved to this file to preserve the overall dependency structure.

File: src/types/api.js [hunk 19]
```diff
@@ -83,9 +83,10 @@
  * TODO - Fill out remaining properties
  *
  * @typedef Profile
- * @prop {string} userid
+ * @prop {string|null} userid
  * @prop {Object} preferences
- * @prop {boolean} preferences.show_sidebar_tutorial
+ * @prop {boolean} [preferences.show_sidebar_tutorial]
+ * @prop {Object} features
```
Comment (robertknight) [inline src/types/api.js:89]:
```suggestion
 * @prop {Object.<string, boolean>} features
```

IIRC `features` maps feature flag names to whether they are enabled or not.

File: src/sidebar/store/modules/annotations.js (hunk 10 above)
Comment (LMS007) [reply src/sidebar/store/modules/annotations.js:558]: installed, thanks!

File: src/sidebar/store/modules/defaults.js (hunk 11 above)
Comment (LMS007) [reply src/sidebar/store/modules/defaults.js:66]: yep, that was my question above https://github.com/hypothesis/client/pull/2370#discussion_r459461194

this makes more sense :)

File: src/sidebar/store/modules/links.js (hunk 15 above)
Comment (LMS007) [reply src/sidebar/store/modules/links.js:29]: I think its a good time to do this in fact

File: src/sidebar/store/modules/real-time-updates.js (hunk 16 above)
Comment (LMS007) [reply src/sidebar/store/modules/real-time-updates.js:186]: agreed

File: src/sidebar/store/modules/drafts.js (hunk 14 above)
Comment (LMS007) [reply src/sidebar/store/modules/drafts.js:203]: I'm not sure I fully understand the danger here. It appears it's actually return `Draft[]]`, perhaps cast to an `Annotation`, but it's returning data from the drafts store nevertheless. Are you implying there is a linkage between copies in the drafts store and annotations store that could potentially get out of sync? If that's the case we may need to redesign the drafts/annotations stores and set up a different type of relationship.

For now, I'm going to go ahead and change this type to {Draft[]}, we can discuss further on Monday.

File: src/sidebar/store/modules/toast-messages.js (hunk 18 above)
Comment (LMS007) [reply src/sidebar/store/modules/toast-messages.js:3]: roger that

File: src/sidebar/store/modules/annotations.js [hunk 20]
```diff
@@ -545,6 +543,35 @@ function savedAnnotations(state) {
...
+ * @prop {typeof updateFlagStatus} updateFlagStatus
+ 
+ *
+ * // Selectors
+ * @prop {() => number} annotationCount
+ * @prop {(id: string) => boolean} annotationExists
+ * @prop {(id: string) => Annotation} findAnnotationByID
+ * @prop {(tags: string[]) => string[]} findIDsForTags
+ * @prop {() => string[]} focusedAnnotations
+ * @prop {() => string[]} highlightedAnnotations
```
Comment (robertknight) [inline src/sidebar/store/modules/annotations.js:566]: The fact that `newAnnotations` returns `Annotation`s whereas `focusedAnnotations` returns IDs is surprising, but this PR is just surfacing an existing issue.

File: src/sidebar/store/modules/annotations.js [hunk 21]
```diff
@@ -545,6 +543,35 @@ function savedAnnotations(state) {
...
+ * @prop {() => boolean} isWaitingToAnchorAnnotations
+ * @prop {() => Annotation[]} newAnnotations
+ * @prop {() => Annotation[]} newHighlights
+ * @prop {() => boolean} noteCount
+ * @prop {() => boolean} orphanCount
```
Comment (robertknight) [inline src/sidebar/store/modules/annotations.js:571]: `noteCount` and `orphanCount` return numbers, not boolean.

File: src/sidebar/store/modules/annotations.js (hunk 20 above)
Comment (LMS007) [reply src/sidebar/store/modules/annotations.js:566]: indeed, I'll do a follow up refactor on that
